### PR TITLE
Increase Script Class Name access.

### LINKED ISSUE #4708: Ability to get class name string from Reference
### Describe the project you are working on

General GDScript improvements.

### Describe the problem or limitation you are having in your project

For now it's not possible to get class name string from Reference object out of box.

I fount it's already has been discussed in Q/A:
https://godotengine.org/qa/50783/get-class-name-as-string-from-class-reference
https://godotengine.org/qa/46057/how-to-get-the-class-name-of-a-custom-node

But still not implemented.

Of course it can be workarounded with custom variable or function. But it's would be much more convenient if it's will be awailable out of box.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Described above.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

For example trough function get_class()
```gdscript
class MyClass:
	pass

class MyChildClass extends MyClass:
	pass

func _ready():
	print(MyClass.new().get_class()) # will print out: MyClass
	print(MyChildClass.new().get_class()) # will print out: MyChildClass 
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Yes, it's can be workarounded by custom variable like this:
```gdscript
class MyClass:
	var _class_name: String = 'MyClass'

class MyChildClass extends MyClass:
	func _init():
		_class_name = 'MyChildClass'
```

or by curom function like this:
```gdscript
class MyClass:
	func get_class() -> String:
		return 'MyClass'

class MyChildClass extends MyClass:
	func get_class() -> String:
		return 'MyChildClass'
```

But that is tedious enough.

### Is there a reason why this should be core and not an add-on in the asset library?

Beacause it's a GDScript core improvement

### LINKED ISSUE #21789: get_class() and is_class() not returning class_name
**v3.1.alpha.custom_build.5307043**

The new `class_name` keyword doesn't affect the results of the methods for `get_class()` and `is_class()`.

I noticed this earlier when encountering https://github.com/godotengine/godot/issues/21461, and it feels like maybe it might have been overlooked.  Not sure if it falls under bug or a feature request, or if maybe its even intentional and required.

It would be very useful to have these methods return the custom class types in certain situations.

### PR DESCRIPTION
**Pull request by Diddykonga**
Added `StringName ScriptServer::get_global_class_name(StringName script_path)`
Added `StringName Object::get_script_class_name()`
Added `StringName Script::get_global_class_name()`
Changed `*::is_class()` to check Script Class Names as well. (Supports Script Inheritance)

Fixes: https://github.com/godotengine/godot/issues/21789
Might Fix: https://github.com/godotengine/godot-proposals/issues/4708

Test Project: [ClassNameTest.zip](https://github.com/godotengine/godot/files/8948213/ClassNameTest.zip)

### CODE REVIEW

File: core/object/script_language.cpp
Comment (willnationsdev) [inline core/object/script_language.cpp:241]: If the code is running in the editor, then it should use `EditorNode::get_editor_data().script_class_get_name(path)`) (something like that). EditorData maintains a reverse-lookup map to speed up these kinds of queries. Should be able to use `Engine::get_singleton()->is_editor_hint()` and the `#ifdef TOOLS_ENABLED` macro to help manipulate the contents of this method correctly.

File: core/object/object.h
Comment (willnationsdev) [inline core/object/object.h:713]: While I believe that it would clearly make more *sense* for an `is_class` method to return the name of the best understood typename, as you have implemented, I also believe that modifying this method in this manner is too dangerous. `is_class()` is used frequently throughout the engine with the assumption that the returned class name will correspond to an actual natively-implemented class stored in the `ClassDB`. This makes it still work properly and be applicable for extensions, but not so for scripted types.

Rather than this sort of solution, I would *much* prefer to see something where we add an entirely different method to the base class that handles abstracting away the complexity of the script class name vs. the native class name. In which case, we should decide on the name of that method in the related Issue (#21789) probably. Other than that though, I like a lot of the modifications you've made, personally.

This also goes for the `is_class` method definition in the object macro further up in this same file.

**Review by willnationsdev — CHANGES_REQUESTED** [review]
I don't have too much authority, but I think that, for the most part, you've got a nice set of changes here. There's just a little tweaking necessary to really make it compatible with the rest of the codebase without forcing huge portions of the engine to be refactored and retested.

File: core/object/object.h
Comment (Diddykonga) [reply core/object/object.h:713]: I'm confused as to why adding the Script type checking to `is_class` could be a bad thing, or why a separate method would be required? This method will still work if passed a Native or Extension type.

The only possible downside/regression I can see is if a developer were to shadow an Native or Extension type with a Script type, that false-positives could happen, but shadowing is bad practice anyways.

Thank you. :)

File: core/object/script_language.cpp
Comment (Diddykonga) [reply core/object/script_language.cpp:241]: I'm unsure if adding this would make sense, It's my understanding the Editor should be separate from Core, but I believe I have seen these checks before in Core code, though I might be wrong, so I dont know lul.

Comment (willnationsdev) [reply core/object/script_language.cpp:241]: Yeah, I mentioned that they probably wouldn't want tool references in the core code in my edit. It's probably fine.

File: core/object/object.h
Comment (willnationsdev) [reply core/object/object.h:713]: > I'm confused as to why adding the Script type checking to is_class could be a bad thing

I think I initially confused this with `get_class()` for some reason which is used pretty extensively in the engine source code. I don't know how often `is_class()` is used, but if it's a lot less then *maybe* it would be less problematic? Still would end up adding to the inconsistency vs. the similarly named `get_class()` though.

It's less so that it's a "bad thing" if you added this functionality, but more so that you'd need to double check every reference to this method in the rest of the codebase to ensure it's logic is still consistent should the value return `true` unexpectedly. For example, if the code passed in a typename string to verify that the given typename is the name of an inherited *native* type when (cause it needs to verify it to do some native-only operation that is incompatible with scripted types), then making this change would suddenly make it possible for previously unsupported - and correctly incompatible - types to now slip through that check. That's the "risk". We would need to ensure that things that use this method to control flow logic don't e.g. do anything with the `ClassDB` since those operations would surely fail and/or log an error after this change is made.

> why a separate method would be required?

If a separate method is defined instead, then it means no one would have to do the above "cleanup" refactoring/investigating throughout the rest of the codebase after the change is made. BUT, I do agree that it could lead to redundancy / choice paralysis or confusion regarding which method should be used in which circumstances (a reoccurring source of confusion in Godot's API for newbies imo). So, it's not that it would be required. Just that IF we made the change, we'd need a corresponding correctional sweep of the rest of the codebase to go alongside it.

Comment (Diddykonga) [reply core/object/object.h:713]: >1st part
I can understand if this could cause some problems in places with Shadowing, but honestly I think this should be fixed on the GDScript side, in that `class_name` shouldn't be able to contain the name of an already existing Type. At least that seems like the more logical change personally anyways.
>2nd part
Instead of a separate method, I would instead offer a defaulted second parameter as to whether or not Script Types are considered.

Comment (willnationsdev) [reply core/object/object.h:713]: That sounds like a reasonable approach.

Comment (Diddykonga) [reply core/object/object.h:713]: > That sounds like a reasonable approach.

Okay, well I have a commit archived that I can push anytime before this gets merged if that is the route we wish to take.
However personally I would rather the Shadowing get fixed where it is being caused, so that the second parameter is not necessary , although that would need to be a separate PR to this one.

Comment (Diddykonga) [reply core/object/object.h:713]: So I just checked: https://github.com/godotengine/godot/blob/7686cfec64a9cbfe1c27bfa4f8b1eaf43755830e/modules/gdscript/gdscript_analyzer.cpp#L226

GDScript already checks for Shadowing and throws errors, which means I don't see a reason for this as is, to cause any issues.

**Comment by Diddykonga** [comment]
Just remembered because I'm dumb, that `Script` is also an `Object`, meaning it doesn't require it's own method lul
Updating now, should be good after that.

**Comment by YuriSizov** [comment]
Note that we now have a method in ProjectSettings to get a list of global classes, so something about this PR should likely be changed.

**Comment by Diddykonga** [comment]
After reviewing what has changed, and also the state of extensions, I believe this PR is no longer needed.

`Script::get_global_name()` just needs to be exposed to GDScript.

**Comment by API-Beast** [comment]
> After reviewing what has changed, and also the state of extensions, I believe this PR is no longer needed.
> 
> `Script::get_global_name()` just needs to be exposed to GDScript.

As of January 2024, Script::get_global_name is still not exposed.

**Comment by AThousandShips** [comment]
Yes [it is](https://docs.godotengine.org/en/latest/classes/class_script.html#class-script-method-get-global-name)

**Comment by akien-mga** [comment]
It was exposed by #80487, which will be in 4.3. I assume @API-Beast tested on 4.2.x.

**Pull request closed without merging**
